python/glow/gwas/linear_regression.py

```diff
@@ -0,0 +1,149 @@
+        A Spark DataFrame that contains:
+        - All columns from `genotype_df` except the `values_column`
+        - `effect`: The effect size estimate for the genotype
+        - `stderror`: The estimated standard error
+        - `tvalue`: The T statistic
+        - `pvalue`: P value estimed from a two sided t-test
+        - `phenotype`: The phenotype name as determined by the column names of `phenotype_df`
+    '''
+
+    # Validate input
+    for col in covariate_df:
+        __assert_all_present(covariate_df, col, 'covariate')
+    if not covariate_df.empty:
+        if phenotype_df.shape[0] != covariate_df.shape[0]:
+            raise ValueError(
+                f'phenotype_df and covariate_df must have the same number of rows ({phenotype_df.shape[0]} != {covariate_df.shape[0]}'
+            )
+
+    # Construct output schema
+    result_fields = [
+        StructField('effect', DoubleType()),
+        StructField('stderror', DoubleType()),
+        StructField('tvalue', DoubleType()),
+        StructField('pvalue', DoubleType()),
+        StructField('phenotype', StringType())
+    ]
+    fields = [field
+              for field in genotype_df.schema.fields if field.name != values_column] + result_fields
+    result_struct = StructType(fields)
+
+    C = covariate_df.to_numpy(np.float64, copy=True)
+    if fit_intercept:
+        intercept = np.ones((phenotype_df.shape[0], 1))
+        C = np.hstack((intercept, C)) if C.size else intercept
+
+    # Prepare covariate basis and phenotype residuals
+    Y = phenotype_df.to_numpy(np.float64, copy=True)
+    Y_mask = (~np.isnan(Y)).astype(np.float64)
+    np.nan_to_num(Y, copy=False)
+    Q = np.linalg.qr(C)[0]
+    Y = _residualize_in_place(Y, Q) * Y_mask
+    YdotY = np.sum(Y * Y, axis=0)
+    dof = C.shape[0] - C.shape[1] - 1
+
+    def map_func(pdf_iterator):
+        for pdf in pdf_iterator:
+            yield _linear_regression_inner(pdf, Y, YdotY, Y_mask, Q, dof,
+                                           phenotype_df.columns.to_series().astype('str'))
+
+    return genotype_df.mapInPandas(map_func, result_struct)
+
+
+@typechecked
+def _add_intercept(C: NDArray[(Any, Any), np.float64], num_samples: int):
+    intercept = np.ones((num_samples, 1))
+    return np.hstack((intercept, C)) if C.size else intercept
+
+
+@typechecked
+def _einsum(subscripts: str, *operands: NDArray):
+    '''
+    A wrapper around np.einsum to ensure uniform options.
+    '''
+    return np.einsum(subscripts, *operands, casting='no')
+
+
+@typechecked
+def _residualize_in_place(M: NDArray[(Any, Any), np.float64],
+                          Q: NDArray[(Any, Any), np.float64]) -> NDArray[(Any, Any), np.float64]:
+    '''
+    Residualize a matrix in place using an orthonormal basis. The residualized matrix
+    is returned for easy chaining.
+    '''
+    M -= Q @ (Q.T @ M)
+    return M
+
+
+@typechecked
+def _linear_regression_inner(genotype_df: pd.DataFrame, Y: NDArray[(Any, Any), np.float64],
```

Comment:
Can you use a name other than `genotype_df` here as it is used for the spark dataframe before?